### PR TITLE
Replaced implementation() by api() for opensearch libs

### LINKED ISSUE #609: [BUG] Transitive Dependencies are not exposed to dependent projects
### What is the bug?

I am working on a "from scratch" setup.  I have declared a dependency only to `opensearch-sdk-java` 2.0.0-SNAPSHOT.

When attempting to import a class in `opensearch` 3.0.0-SNAPSHOT package `org.opensearch.extensions`, compilation fails.

### How can one reproduce the bug?

1. Create a new project that depends only on `opensearch-sdk-java`, like [this one](https://github.com/dbwiddis/CRUDExtension)
2. Import a class from a transitive dependency such as OpenSearch.
3. Observe compile time failure

Dependency resolution shows the import is `runtime` only:
```
$ mvn dependency:tree
[INFO] CRUDExtension:CRUDExtension:jar:0.0.1-SNAPSHOT
[INFO] \- org.opensearch.sdk:opensearch-sdk-java:jar:2.0.0-SNAPSHOT:compile
[INFO]    +- org.opensearch:opensearch:jar:3.0.0-SNAPSHOT:runtime
```

Adding the transitive dependency manually resolves this problem, but should not be required.

### What is the expected behavior?

Developers should not be required to declare dependencies themselves; transitive compile dependencies should resolve automatically.

### Do you have any additional context?

According to [Gradle Docs](https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_separation)
> The key difference between the standard Java plugin and the Java Library plugin is that the latter introduces the concept of an API exposed to consumers. A library is a Java component meant to be consumed by other components. It’s a very common use case in multi-project builds, but also as soon as you have external dependencies.
>
> The plugin exposes two [configurations](https://docs.gradle.org/current/userguide/declaring_dependencies.html#sec:what-are-dependency-configurations) that can be used to declare dependencies: api and implementation. The api configuration should be used to declare dependencies which are exported by the library API, whereas the implementation configuration should be used to declare dependencies which are internal to the component.

This suggests that this line in `build.gradle`:

```groovy
implementation("org.opensearch:opensearch:${opensearchVersion}")
```

should be:

```groovy
api("org.opensearch:opensearch:${opensearchVersion}")
```

This is likely true for other transitive dependencies such as `log4j-api` and the OpenSearch clients.

### PR DESCRIPTION
**Pull request by nassipkali**
### Description
Replaced implementation() by api() for opensearch libs

### Issues Resolved
Fixes #609 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by nassipkali** [comment]
@dbwiddis I tested it on CRUDExtension and it removed requirement of defining opensearch, clients in dependencies manually. But still needs add to dependencies a lot of apache-lucene libraries

**Comment by nassipkali** [comment]
@dbwiddis Like this:
```xml
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-core</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-analysis-common</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-backward-codecs</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-grouping</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-highlighter</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-join</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-memory</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-misc</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-queries</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-queryparser</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-sandbox</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-spatial-extras</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-spatial3d</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
<dependency>
	<groupId>org.apache.lucene</groupId>
	<artifactId>lucene-suggest</artifactId>
	<version>9.6.0-SNAPSHOT</version>
</dependency>
```

File: build.gradle [hunk 1]
```diff
@@ -90,13 +90,13 @@ dependencies {
...
+    api("org.opensearch:opensearch:${opensearchVersion}")
     implementation("org.apache.logging.log4j:log4j-api:${log4jVersion}")
     implementation("org.apache.logging.log4j:log4j-core:${log4jVersion}")
     implementation("org.apache.logging.log4j:log4j-jul:${log4jVersion}")
-    implementation("org.opensearch.client:opensearch-rest-high-level-client:${opensearchVersion}")
-    implementation("org.opensearch.client:opensearch-rest-client:${opensearchVersion}")
-    implementation("org.opensearch.client:opensearch-java:${opensearchVersion}")
+    api("org.opensearch.client:opensearch-rest-high-level-client:${opensearchVersion}")
+    api("org.opensearch.client:opensearch-rest-client:${opensearchVersion}")
+    api("org.opensearch.client:opensearch-java:${opensearchVersion}")
```
Comment (dbwiddis) [inline build.gradle:99]: As a note to other reviewers, there are open issues to remove these clients completely: #612 and #732.  So this is appropriate temporarily but I hope eventually these lines disappear entirely.

**Review by dbwiddis — CHANGES_REQUESTED** [review]
See below.

**Comment by dbwiddis** [comment]
> @dbwiddis I tested it on CRUDExtension and it removed requirement of defining opensearch, clients in dependencies manually. But still needs add to dependencies a lot of apache-lucene libraries

Those libraries should come in via OpenSearch core, though, not the RestHighLevelClient... I think?

**Comment by nassipkali** [comment]
@dbwiddis What does Validate Gradle Wrapper check? After last commit that check failed

**Review by dbwiddis — APPROVED** [review]
LGTM!

File: build.gradle [hunk 2]
```diff
@@ -13,7 +13,7 @@ import java.nio.file.Files
 import org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestFramework
 
 plugins {
-    id 'java'
+    id 'java-library'
```
Comment (owaiskazi19) [inline build.gradle:16]: Any specific reason to change the plugin here?

Comment (nassipkali) [reply build.gradle:16]: > Any specific reason to change the plugin here?

api() doesn't work if java-library not imported.

File: build.gradle [hunk 3]
```diff
@@ -90,13 +90,13 @@ dependencies {
     def junitPlatform = "1.9.3"
     def jaxbVersion = "2.3.1"
 
-    implementation("org.opensearch:opensearch:${opensearchVersion}")
+    api("org.opensearch:opensearch:${opensearchVersion}")
```
Comment (owaiskazi19) [inline build.gradle:93]: `api` tends to make the internal libraries as public whereas `implementation` won't let us access the internal libraries by considering them `private`. I don't understand the reason for this change.

Comment (owaiskazi19) [reply build.gradle:93]: Okay. I see #609

File: build.gradle (hunk 2 above)
Comment (dbwiddis) [reply build.gradle:16]: `java` is actually deprecated, `java-library` is the new thing if you have a project intended to be consumed by other projects (which the SDK does).

**Review by owaiskazi19 — APPROVED** [review]

**Pull request merged**